src/query/src/window_sort.rs

```diff
@@ -0,0 +1,3101 @@
+    }
+
+    /// make `in_progress` as a new `DfSendableRecordBatchStream` and put into `sorted_input_runs`
+    fn build_sorted_stream(&mut self) -> datafusion_common::Result<()> {
+        if self.in_progress.is_empty() {
+            return Ok(());
+        }
+        let data = std::mem::take(&mut self.in_progress);
+
+        let new_stream = MemoryStream::try_new(data, self.schema(), None)?;
+        self.sorted_input_runs.push(Box::pin(new_stream));
+        Ok(())
+    }
+
+    /// Start merging sort the current working set
+    fn start_new_merge_sort(&mut self) -> datafusion_common::Result<()> {
+        if !self.in_progress.is_empty() {
+            return internal_err!("Starting a merge sort when in_progress is not empty")?;
+        }
+
+        self.set_next_working_range();
+
+        let streams = std::mem::take(&mut self.sorted_input_runs);
+        if streams.is_empty() {
+            return Ok(());
+        } else if streams.len() == 1 {
+            self.merge_stream
+                .push_back(streams.into_iter().next().unwrap());
+            return Ok(());
+        }
+
+        let fetch = self.remaining_fetch();
+        let reservation = MemoryConsumer::new(format!("WindowedSortStream[{}]", self.merge_count))
+            .register(&self.memory_pool);
+        self.merge_count += 1;
+
+        let resulting_stream = streaming_merge(
+            streams,
+            self.schema(),
+            &[self.expression.clone()],
+            BaselineMetrics::new(&self.metrics, 0),
+            self.batch_size,
+            fetch,
+            reservation,
+        )?;
+        self.merge_stream.push_back(resulting_stream);
+        // this working range is done, move to next working range
+        Ok(())
+    }
+
+    /// Remaining number of rows to fetch, if no fetch limit, return None
+    /// if fetch limit is reached, return Some(0)
+    fn remaining_fetch(&self) -> Option<usize> {
+        let total_now = self.produced;
+        self.fetch.map(|p| p.saturating_sub(total_now))
+    }
+}
+
+impl Stream for WindowedSortStream {
+    type Item = datafusion_common::Result<DfRecordBatch>;
+
+    fn poll_next(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<datafusion_common::Result<DfRecordBatch>>> {
+        self.poll_next_inner(cx)
+    }
+}
+
+impl RecordBatchStream for WindowedSortStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+/// split batch to sorted runs
+fn split_batch_to_sorted_run(
+    batch: DfRecordBatch,
+    expression: &PhysicalSortExpr,
+) -> datafusion_common::Result<SortedRunSet<Timestamp>> {
+    // split input rb to sorted runs
+    let sort_column = expression.evaluate_to_sort_column(&batch)?;
+    let sorted_runs_offset = get_sorted_runs(sort_column.clone())?;
+    if let Some(run) = sorted_runs_offset.first()
+        && run.offset == 0
+        && run.len == batch.num_rows()
+        && sorted_runs_offset.len() == 1
```

Comment:
if `sorted_runs_offset.len() == 1`, it must satisfy `run.offset == 0` and `run.len == batch.num_rows()`. Maybe we should check this.